p2p/dialer.go

```diff
@@ -0,0 +1,191 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package p2p
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/mysteriumnetwork/node/communication/nats"
+	"github.com/mysteriumnetwork/node/core/ip"
+	"github.com/mysteriumnetwork/node/core/port"
+	"github.com/mysteriumnetwork/node/identity"
+	"github.com/mysteriumnetwork/node/pb"
+	"github.com/rs/zerolog/log"
+	"google.golang.org/protobuf/proto"
+)
+
+// Dialer knows how to exchange p2p keys and encrypted configuration and creates ready to use p2p channels.
+type Dialer interface {
+	// Dial exchanges p2p configuration via broker, performs NAT pinging if needed
+	// and create p2p channel which is ready for communication.
+	Dial(consumerID, providerID identity.Identity, serviceType string, timeout time.Duration) (Channel, error)
+}
+
+// NewDialer creates new p2p communication dialer which is used on consumer side.
+func NewDialer(broker brokerConnector, address string, signer identity.SignerFactory, verifier identity.Verifier, ipResolver ip.Resolver, consumerPinger natConsumerPinger) Dialer {
+	return &dialer{
+		broker:         broker,
+		brokerAddress:  address,
+		ipResolver:     ipResolver,
+		signer:         signer,
+		verifier:       verifier,
+		portPool:       port.NewPool(),
+		consumerPinger: consumerPinger,
+	}
+}
+
+// dialer implements Dialer interface.
+type dialer struct {
+	portPool       *port.Pool
+	broker         brokerConnector
+	consumerPinger natConsumerPinger
+	signer         identity.SignerFactory
+	verifier       identity.Verifier
+	ipResolver     ip.Resolver
+	brokerAddress  string
+}
+
+// Dial exchanges p2p configuration via broker, performs NAT pinging if needed
+// and create p2p channel which is ready for communication.
+func (m *dialer) Dial(consumerID, providerID identity.Identity, serviceType string, timeout time.Duration) (Channel, error) {
+	config, err := m.exchangeConfig(consumerID, providerID, serviceType, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("could not exchange config: %w", err)
+	}
+
+	var remotePort, localPort int
+	var conn0 *net.UDPConn
+	var serviceConn *net.UDPConn
+	if len(config.peerPorts) == 1 {
+		localPort = config.localPorts[0]
+		remotePort = config.peerPorts[0]
+	} else {
+		log.Debug().Msgf("Pinging provider %s with public IP %s using ports %v:%v", providerID.Address, config.peerPublicIP, config.localPorts, config.peerPorts)
+		conns, err := m.consumerPinger.PingProviderPeer(config.peerPublicIP, config.localPorts, config.peerPorts, consumerInitialTTL, requiredConnLen)
+		if err != nil {
+			return nil, fmt.Errorf("could not ping peer: %w", err)
+		}
+		conn0 = conns[0]
+		localPort = conn0.LocalAddr().(*net.UDPAddr).Port
+		remotePort = conn0.RemoteAddr().(*net.UDPAddr).Port
+		serviceConn = conns[1]
```

Comment:
This logic assumes that there will be only 1 service in setup. This looks like a hack since we are passing `requiredConnLen` which might be larger than 2. In reality we should not limit in setup service amount to 1.

Would leave it like this for now, but would create an issue, for further improvement.